### PR TITLE
Add representation profiling under input changes

### LINKED ISSUE #3626: [Feature]: Model Behavior Analysis under Kornia Augmentations (kornia.contrib)
### 🚀 Feature Description

Kornia provides powerful differentiable augmentations, but there is currently no built-in tooling to analyze how vision models behave internally when exposed to these augmentations.

I propose adding a small diagnostic utility that applies Kornia augmentations to an input image and analyzes how intermediate model representations change across layers. The goal is to measure layer sensitivity and representation stability under augmentation-induced perturbations.

The feature would primarily target research and debugging workflows and could live under kornia.contrib as experimental tooling.

### 📂 Feature Category

Augmentation

### 💡 Motivation

When training or evaluating vision models with augmentations, it is often unclear how those perturbations affect internal representations of the network. If model performance changes under augmentations, it is difficult to determine which layers are sensitive to the perturbations and whether the learned representations remain stable.

Currently, Kornia provides powerful augmentation operators, but there is no direct way within the ecosystem to inspect how models respond internally to those transformations.

Providing a lightweight analysis utility could help users better understand representation robustness, debug model behavior under augmentations, and study invariance properties of vision models.

### 💭 Proposed Solution

One possible approach would be to implement a small analysis utility that applies Kornia augmentations to an input image and measures how model representations change under those perturbations.

The workflow could look roughly as follows:

image
↓
Kornia augmentation pipeline
↓
model forward pass
↓
capture intermediate layer activations
↓
compute layer sensitivity scores
↓
analyze representation stability for selected layers
↓
evaluate robustness of the final embedding

Intermediate activations could be captured using PyTorch forward hooks, allowing the tool to remain model-agnostic and compatible with common PyTorch vision architectures.

Since this functionality is primarily intended for analysis and experimentation rather than core vision operators, it could potentially live under kornia.contrib.

### 🔄 Alternatives Considered

The same type of analysis can technically be performed using ad-hoc scripts by applying augmentations and manually inspecting intermediate activations. However, this approach usually requires implementing forward hooks, feature comparison logic, and analysis pipelines from scratch for each experiment.

While there are external robustness and interpretability libraries, they are generally designed around their own augmentation frameworks and do not integrate naturally with Kornia’s differentiable augmentation pipeline.

Providing a small utility within the Kornia ecosystem would make this type of representation analysis easier to perform while keeping the workflow fully compatible with Kornia-based augmentation pipelines.

### 🎯 Use Cases

• Robustness analysis – Evaluate how sensitive different layers of a vision model are to common augmentations such as rotations, blur, or brightness changes.

• Model debugging – Identify layers where representations change significantly under small perturbations, which can help diagnose instability or brittleness in trained models.

• Representation analysis – Study how internal feature representations evolve across layers when inputs are perturbed using Kornia augmentations.

• Research workflows – Support experiments related to invariance, robustness, and representation stability in vision models using Kornia-based augmentation pipelines.

### 📝 Additional Context

The analysis would rely on PyTorch forward hooks to capture intermediate activations and would remain model-agnostic so it can work with common PyTorch vision architectures.

The augmentation stage would leverage existing Kornia augmentation pipelines (e.g., AugmentationSequential), allowing users to analyze model behavior under the same transformations used during training or evaluation.

Since this functionality is primarily intended for experimentation and diagnostics rather than core operators, it would likely fit best under kornia.contrib.

### 🤝 Contribution Intent

- [x] I plan to submit a PR to implement this feature
- [ ] I'm requesting this feature but not planning to implement it

### PR DESCRIPTION
**Pull request by navya7821**
## 📝 Description

Fixes #3626

---

## 🛠️ Changes Made
- Added `kornia.profiler` module for representation-level analysis
- Implemented `ModelProfiler` with context-manager API for layer-wise feature extraction
- Added generic API `model_profile_under_input_changes` for comparing representations across inputs
- Added config-driven augmentation wrapper built on top of the core API
- Implemented modular metrics:
  - cosine similarity
  - linear CKA
  - normalized L2
  - gram similarity

---

## 🧪 How Was This Tested?

- [ ] **Unit Tests:** Not added (can be added based on maintainer feedback)
- [x] **Manual Verification:** 
  - Verified feature extraction across multiple layers
  - Tested input pairing and group-based comparison
  - Validated metric outputs across different inputs and augmentations
- [x] **Performance/Edge Cases:**
  - Handles mismatched group sizes (with safe truncation)
  - Supports non-4D tensors (Gram metric safely skipped)
  - Works with arbitrary models and input structures

---

## 🕵️ AI Usage Disclosure

- [x] 🟡 **AI-assisted:** I used AI for minor formatting and refinement.

---

## 🚦 Checklist

- [x] I am assigned to the linked issue (required before PR submission)
- [x] The linked issue has been approved by a maintainer
- [x] I have performed a self-review of my code
- [x] My code follows the existing style guidelines of this project
- [x] I have commented my code where necessary
- [ ] I have added tests that prove my feature works
- [ ] UI changes

---

## 💭 Additional Context

This implementation follows the design direction discussed in the issue:
- Core functionality is framed as representation comparison under input changes
- Augmentations are implemented as a thin wrapper on top of the generic API
- Metrics are modularized (one per file)

Higher-level composite metrics (e.g., sensitivity/stability scoring) were intentionally excluded to keep the core minimal and composable. These can be layered on top if needed.

### CODE REVIEW

**Comment by navya7821** [comment]
Hi @shijianjian, I wanted to follow up on this PR. Based on your earlier suggestions ( from the issue #3626), I’ve been considering aligning the implementation more closely with a minimal, library-friendly profiler (focusing on feature extraction and grouping, and keeping analysis components external).

I’m also planning to base my GSoC proposal on this direction. Does this scope align with what you’d consider appropriate for Kornia, or would you recommend narrowing it further?

Happy to adjust the PR accordingly.

**Comment by shijianjian** [comment]
Hi @navya7821, the GSoC focuses on [kornia-rs](https://github.com/kornia/kornia-rs/wiki/%5B2026%5D-Google-Sumer-of-Code-Application), not on Kornia-py. You may check on the project list there.

Kornia-rs focuses more on the IoT and robotics side. The representation profiling is more on the research side than deployment. I do not think there is a need for representation analysis for embeded hardware yet.

**Comment by navya7821** [comment]
Thank you @shijianjian  for the clarification regarding GSoC and kornia-rs.

For this PR, I’d like to continue refining it for Kornia-Py. Based on your earlier feedback, I’m planning to scope it down to a minimal, library-friendly profiler focused on feature extraction and grouping, and keep any analysis components external.

Would this direction be acceptable for Kornia-Py, or would you recommend further narrowing or restructuring the PR?

Happy to iterate accordingly.

File: kornia/profiler/metrics/cosine.py [hunk 1]
```diff
@@ -13,21 +13,23 @@
...
 # limitations under the License.
-# limitations under the License.
-import sys
-from pathlib import Path
+#
+
+import torch.nn.functional as F
+
+
+def cosine_similarity(f1, f2):
```
Comment (shijianjian) [inline kornia/profiler/metrics/cosine.py]: needs typing

File: kornia/profiler/augment.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    "rotation": lambda cfg: K.RandomRotation(degrees=cfg.get("degrees", 15.0), p=1.0),
+    "blur": lambda cfg: K.RandomGaussianBlur(
+        kernel_size=cfg.get("kernel_size", (3, 3)),
+        sigma=cfg.get("sigma", (0.1, 2.0)),
+        p=1.0,
+    ),
+    "brightness": lambda cfg: K.RandomBrightness(
+        brightness=cfg.get("brightness", (0.8, 1.2)),
+        p=1.0,
+    ),
```
Comment (shijianjian) [inline kornia/profiler/augment.py]: avoid lambda. This redefines a module on every call

**Review by shijianjian — COMMENTED** [review]
It looks good in general. Please add typing and more docs for each function.

You may also need to update https://github.com/kornia/kornia/tree/main/docs/source

File: kornia/profiler/metrics/cosine.py (hunk 1 above)
Comment (navya7821) [reply kornia/profiler/metrics/cosine.py]: Thanks for the feedback!

I’ve added type annotations and improved the docstrings for all functions. I also fixed lint issues raised by pre-commit.

Please let me know if anything else needs adjustment.

File: kornia/profiler/augment.py (hunk 2 above)
Comment (navya7821) [reply kornia/profiler/augment.py]: Good point, thanks. 
I’ve replaced the lambda-based registry with explicit builder functions and added docstrings

**Comment by navya7821** [comment]
Thanks a lot @shijianjian  for the feedback!
I’ve addressed the requested changes (typing, docstrings, and replaced the lambda-based registry). All checks are passing now.

Please let me know if anything else is needed.

File: kornia/profiler/augment.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        brightness=cfg.get("brightness", (0.8, 1.2)),
+        p=1.0,
+    )
+
+
+AUGMENTATION_REGISTRY = {
+    "rotation": _build_rotation,
+    "blur": _build_blur,
+    "brightness": _build_brightness,
+}
```
Comment (shijianjian) [inline kornia/profiler/augment.py:70]: This registry is too stupid. Can you make it smarter with the `inspection` module?

The cfg input can directly input the function signature + argumetns

File: kornia/profiler/model_profiler.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        Returns:
+            The current ``ModelProfiler`` instance.
+        """
+        self.extractor = FeatureExtractor(self.model, self.layers, processing=self.processing)
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        r"""Exit context manager and remove registered hooks."""
+        if hasattr(self.extractor, "remove_hooks"):
+            self.extractor.remove_hooks()
```
Comment (shijianjian) [inline kornia/profiler/model_profiler.py:82]: Add tests to this

**Review by shijianjian — CHANGES_REQUESTED** [review]
Hi, there are many issues: 1) typing is not added everywhere, 2) missing tests for hook registration cleanup, metric correctness, multiple forward calls, CUDA tensors, etc., 3) file names such as api.py are too generic.

Meanwhile, please write some small docs for the support of various models with different inputs, such as
```
model(image)
model({"image": image, "mask": mask})
model(left, right)
```

We need to know the boundary of your current implementation.

The metrics shall have a modulized API:
```python
class Metric(Protocol):  # maybe the Protocol can simply be a kornia.Module at this stage.
    def __call__(self, x, y) -> Tensor: ...
```


Currently, I feel like the core contribution is the `representation collector`, for representation extraction + comparison. So the module remains generic for future use cases beyond augmentation-based analysis.


For the folder structure, I think
--- profiler
| --- metrics
| --- collectors (your feature extractor is a LayerCollector, you shall implement a collector base class)
| --- engines (your current AugmentationPipeline)
| profiler.py

So, the usage can be:
```python
from kornia.profiler import Profiler

Profiler(model, collector=LayerCollector(layers), pre_metric_transform=Flatten() | "flatten", metrics=[CKA(), "linear"]):
```


Probably, you can also ask ChatGPT/Claude more for this PR. The major contribution I would like to include in this PR is a generic profiler module. And the augmentation profiling is a good starting point.

**Review by edgarriba — COMMENTED** [review]
Thanks again for driving the design discussion on #3626 — the direction is agreed, but this PR still lacks the items from the April review: a test suite (hook registration/cleanup, metric correctness against reference CKA values, CUDA tensors), complete typing on the public functions, the collectors/engines restructure with a `Metric` protocol, and docs pages including the supported input signatures.

Since there's been no activity since April, we'll leave it as changes-requested and let the stale process take it if you can't continue — but happy to pick it back up whenever you can; the core `ModelProfiler` idea is worth landing properly.

**Comment by navya7821** [comment]
Apologies for the long delay here. I’m glad to see that this PR is still under consideration, and I really appreciate the feedback and the opportunity to continue working on it. I’ll be picking this up again shortly and working through the remaining review comments.
